cineast-core/src/main/java/org/vitrivr/cineast/core/features/OCRSearch.java

```diff
@@ -1,24 +1,510 @@
+
+    return spatial_dis + inv_IoU + 10*edit_dis + 0.2*Math.abs(streamFirst.getFirst() - streamLast.getLast());
+  }
+
+  /**
+   * transformString transforms the string by replacing certain characters with others
+   * @param str The string to be transformed
+   * @return The transformed string
+   */
+  private String transformString(String str) { return str.replace("i", "l").replace("l", "1").replace("a", "o");}
+
+  /**
+   * safeText transforms and safes the scene text in the database
+   * @param id id of the shot (the {@link SegmentContainer}) to be processed
+   * @param recognition The scene text that should be safed
+   */
+  private void safeText(String id,String recognition) {
+    if (recognition.equals("")) {
+      return;
+    }
+    this.writer.write(new SimpleFulltextFeatureDescriptor(id, transformString(recognition)));
+  }
+
+  /**
+   * @param coordA
+   * @param coordB
+   * @return the intersection over union of the two rectangular coordinates
+   */
+  private double getIntersectionOverUnion(Quadrilateral_F64 coordA, Quadrilateral_F64 coordB) {
+    double[] boxA = new double[]{coordA.getA().x, coordA.getA().y,coordA.getC().x, coordA.getC().y};
+    double[] boxB = new double[]{coordB.getA().x, coordB.getA().y,coordB.getC().x, coordB.getC().y};
+
+    double xA = Math.max(boxA[0], boxB[0]);
+    double yA = Math.max(boxA[1], boxB[1]);
+    double xB = Math.min(boxA[2], boxB[2]);
+    double yB = Math.min(boxA[3], boxB[3]);
+
+    double interArea = Math.max(0, xB - xA + 1) * Math.max(0, yB - yA + 1);
+
+    double boxAArea = (boxA[2] - boxA[0] + 1) * (boxA[3] - boxA[1] + 1);
+    double boxBArea = (boxB[2] - boxB[0] + 1) * (boxB[3] - boxB[1] + 1);
+
+    return interArea / (boxAArea + boxBArea - interArea);
+  }
+
+  /**
+   * getAverageIntersectionOverUnion takes two lists of coordinates and returns the average IoU
+   * @param coordinates1 Sequentially ordered coordinates (from frame i to frame i+n)
+   * @param coordinates2 Sequentially ordered coordinates (from frame i to frame i+n)
+   * @return The average intersection over union
+   */
+  public double getAverageIntersectionOverUnion(List<Quadrilateral_F64> coordinates1, List<Quadrilateral_F64> coordinates2) {
+    double total_IoU = 0;
+    int count = 0;
+
+    for (int i=0; i<coordinates1.size(); i++) {
+      if (coordinates1.get(i) == null || coordinates2.get(i) == null) {
+        continue;
+      }
+      count++;
+      total_IoU = total_IoU + getIntersectionOverUnion(coordinates1.get(i), coordinates2.get(i));
+    }
+
+    return count==0 ? 0 : total_IoU / count;
+  }
+
+  /**
+   * Extracts the scene text and ingests it using the {@link SimpleFulltextFeatureDescriptor}.
+   * @param shot The {@link SegmentContainer} that should be processed.
+   */
+  @Override
+  public void processSegment(SegmentContainer shot) {
+    TextDetector_EAST detector = new TextDetector_EAST().initialize();
```

Comment:
Can detector / recognizer be re-used? If so, it would be very nice to make them thread-safe and reuse them. Otherwise new instances are created per shot, which is costly for large data sets.